### DISCUSSION TITLE
Question about dict key type and pylance error reporting

### DISCUSSION THREAD

**Discussion by sanderegg** [microsoft/pyright#5426] [General]
Hello,

I am confused by the error I see in Pyright (and actually also mypy) but I am not able to find information as to the reason to it.

considering the following code
```python
class ConstrainedStr(str):
    __slots__ = ()


def func(in_put: dict[str, str]):
    assert in_put  # nosec


var: dict[ConstrainedStr, str] = {}

func(var)
```

I do see an error in pyright such as:
```
Argument of type "dict[ConstrainedStr, str]" cannot be assigned to parameter "in_put" of type "dict[str, str]" in function "func"
  "dict[ConstrainedStr, str]" is incompatible with "dict[str, str]"
    TypeVar "_KT@dict" is invariant
      "ConstrainedStr" is incompatible with "str"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

Although I would have expected that ConstrainedStr being a str itself, it should be acceptable to pass as an argument in a function?
Thanks a lot for some info on that, and maybe how I should fix my code in order to pass the type checker.

**Comment by debonte** [comment]
The key type in a `dict` is invariant, so `dict[ConstrainedStr, str]` is not type compatible with `dict[str, str]` even though `ConstrainedStr` derives from `str`.

There's some additional detail on this behavior in our documentation here: https://microsoft.github.io/pyright/#/type-concepts?id=generic-types.

If you want to treat `var` as a `dict[str, str]`, you can declare it as such when it is first assigned:

```python
class ConstrainedStr(str):
    __slots__ = ()


def func(in_put: dict[str, str]):
    assert in_put  # nosec


var: dict[str, str] = { 
    # Your actual dict entries here
    ConstrainedStr(): "str"
}

func(var)
```

**Reply by sanderegg** [reply]
Hi @debonte, thank you for the valuable answer and the reference. coming from C++ this is a bit weird but I think I kind of got the gist of it.
